### PR TITLE
HPCC-25691 ECLWatch UI to access to ESP daliadmin service

### PR DESCRIPTION
**Pull request by kunalaswani**
Initial DaliAdmin integration into ECLWatch.

Signed-off-by: Kunal Aswani <kunal.aswani@lexisnexis.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [ ] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [ ] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [ ] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: esp/src/src-react/components/DaliAdmin.tsx [hunk 1]
```diff
@@ -0,0 +1,51 @@
+import * as React from "react";
+import { Pivot, PivotItem } from "@fluentui/react";
+import { SizeMe } from "react-sizeme";
+import { pushUrl } from "../util/history";
+import { pivotItemStyle } from "../layouts/pivot";
+import { GetDFSCSV, GetDFSMap, GetDFSParents, GetLogicalFile, GetLogicalFilePart, GetProtectedList, GetValue } from "src/WsDali";
```
Comment (jeclrsg) [inline esp/src/src-react/components/DaliAdmin.tsx:6]: It wasn't available at the time (I think it was only added about a week ago) but you could change this import to be 
`from "@hpcc/comms` and eliminate that src/WsDali file.

File: esp/src/src-react/components/DaliAdmin.tsx [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    tab = "users",
+    name,
+    baseDn
+}) => {
+
+    return <>
+return <SizeMe monitorHeight>{({ size }) =>
+        <Pivot overflowBehavior="menu" style={{ height: "100%" }} selectedKey={tab} onLinkClick={evt => pushUrl(`/wsdali/${tab}/${evt.props.itemKey}`)}>
+            <PivotItem headerText={nlsHPCC.GetDFSCSV} style={pivotItemStyle(size)} >
+                <GetDFSCSV tabname={GetDFSCSV} />
```
Comment (jeclrsg) [inline esp/src/src-react/components/DaliAdmin.tsx:27]: Each of these items, inside of a PivotItem, will be names of other subcomponents that will display the contents of each of those tabs. For example, in [WorkunitDetails](https://github.com/hpcc-systems/HPCC-Platform/blob/master/esp/src/src-react/components/WorkunitDetails.tsx#L42), it imports WorkunitSummary, Variables, etc.

File: esp/src/src-react/components/DaliAdmin.tsx [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+export const DaliAdmin: React.FunctionComponent<DaliAdminProps> = ({
+    filter,
+    tab = "users",
+    name,
+    baseDn
+}) => {
+
+    return <>
+return <SizeMe monitorHeight>{({ size }) =>
```
Comment (jeclrsg) [inline esp/src/src-react/components/DaliAdmin.tsx:24]: This extra "return" could be removed

**Review by jeclrsg — COMMENTED** [review]
@kunalaswani Looking good so far. I see what you meant about the PivotItems. You'll just need to implement some additional components that encapsulate those "pages" for each tab. I linked to the Workunit as an example. And within those components you'll interact with the relevant dali endpoint(s) for getting and setting data.

File: esp/src/src-react/components/DaliAdmin.tsx [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    filter,
+    tab = "GetDFSCSV",
+    name,
+    baseDn
+}) => {
+
+    return <>
+    <SizeMe monitorHeight>{({ size }) =>
+        <Pivot overflowBehavior="menu" style={{ height: "100%" }} selectedKey={tab} onLinkClick={evt => pushUrl(`/wsdali/${tab}/${evt.props.itemKey}`)}>
+            <PivotItem headerText={nlsHPCC.GetDFSCSV} style={pivotItemStyle(size)} >
```
Comment (GordonSmith) [inline esp/src/src-react/components/DaliAdmin.tsx:26]: Each pivot item will need a "key" attribute - which is what the "selectedKey" refers to in the main Pivot

File: esp/src/src/WsDali.ts [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+        retVal.postfix = verArray[1].split("]")[0];
+    }
+    verArray = verArray[0].split("_");
+    if (verArray.length > 1) {
+        retVal.prefix = verArray[0];
+        verArray.splice(0, 1);
+    }
+    retVal.version = verArray.join("_");
+    return retVal;
+}
```
Comment (GordonSmith) [inline esp/src/src/WsDali.ts:84]: This file is not needed as you are using the WsDali service from @hpcc-js/comms

**Review by GordonSmith — CHANGES_REQUESTED** [review]

File: esp/src/src-react/components/GetDFSCSV.tsx [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+    const [dfsCsv, setDfsCsv] = React.useState<string>("");
+
+    React.useEffect(() => {
+        daliService.GetDFSCSV({LogicalNameMask:""}).then(( response ) => {
+            setDfsCsv(response.Result);
+        }).catch(err => logger.error(err));
+    }, []);
+
+    return <div>
+        {dfsCsv}
```
Comment (jeclrsg) [inline esp/src/src-react/components/GetDFSCSV.tsx:27]: I'm not sure what the data looks like coming back from these various requests, but you might use one of the Editor components in .../components/SourceEditor.tsx to display the contents. You can look at .../components/Configuration.tsx for an example of that for XML.

Comment (jeclrsg) [inline esp/src/src-react/components/GetDFSCSV.tsx:27]: I'm not sure what the data looks like coming back from these various requests, but you might use one of the Editor components in .../components/SourceEditor.tsx to display the contents. You can look at .../components/Configuration.tsx for an example of that for XML.

File: esp/src/src-react/components/GetDFSCSV.tsx [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+}
+
+export const GetDFSCSV: React.FunctionComponent<GetDFSCSVProps> = ({
+
+}) => {
+
+    const [dfsCsv, setDfsCsv] = React.useState<string>("");
+
+    React.useEffect(() => {
+        daliService.GetDFSCSV({LogicalNameMask:""}).then(( response ) => {
```
Comment (jeclrsg) [inline esp/src/src-react/components/GetDFSCSV.tsx:21]: And that passed in name prop would go here I think - `...GetDFSCSV({LogicalNameMask: name }).then(...`

Comment (jeclrsg) [inline esp/src/src-react/components/GetDFSCSV.tsx:21]: And that passed in name prop would go here I think - `...GetDFSCSV({LogicalNameMask: name }).then(...`

**Review by jeclrsg — CHANGES_REQUESTED** [review]
@kunalaswani I think I missed your conversation w/ Gordon in the last ECLWatch meeting. So you might've already talked about the things I mentioned.

**Comment by richardkchapman** [comment]
This should target 8.8.x I think?

**Comment by kunalaswani** [comment]
@GordonSmith / @jeclrsg there is an issue with the onchange where after the keystroke it gets stuck in a cycle and doesn't allow additional input rather it rewrites to a blank value. Any thoughts?

File: esp/src/src-react/components/GetDFSCSV.tsx [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+    const [dfsCsv, setDfsCsv] = React.useState<string>("");
+    const [logicalNameMask, setLogicalNameMask] = React.useState<string>("");
+    
+    const onSubmit = React.useCallback(() => { daliService.GetDFSCSV({LogicalNameMask: logicalNameMask}).then(( response ) => {
+        setDfsCsv(response.Result);
+    }).catch(err => logger.error(err)); }, [logicalNameMask]);
+
+    return <div>
+        <TableGroup fields={{
+        "LogicalNameMask": { label: nlsHPCC.LogicalNameMask, type: "string" },
```
Comment (jeclrsg) [inline esp/src/src-react/components/GetDFSCSV.tsx:31]: What's missing here is the value getting passed into the field. You could compare to the TableGroup in [WorkunitSummary.tsx](https://github.com/hpcc-systems/HPCC-Platform/blob/master/esp/src/src-react/components/WorkunitSummary.tsx#L179)

File: esp/src/src-react/components/Menu.tsx [hunk 9]
```diff
@@ -151,6 +152,7 @@ const subMenuItems: SubMenuItems = {
         { headerText: nlsHPCC.Security + " (L)", itemKey: "/security" },
         { headerText: nlsHPCC.Monitoring, itemKey: "/monitoring" },
         { headerText: nlsHPCC.DESDL + " (L)", itemKey: "/desdl" },
+        { headerText: nlsHPCC.DaliAdmin, itemKey: "/daliadmin"},
```
Comment (GordonSmith) [inline esp/src/src-react/components/Menu.tsx:155]: I am trying to get rid of these!  Only needs to be added to new "topology"

File: esp/src/src-react/components/GetValue.tsx [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+    const onSubmit = React.useCallback(() => { return mydaliService.GetValue({ Path: path }).then(( ResultResponse ) => {
+        setValue( ResultResponse.Result );
+    }).catch(err => logger.error(err)); }, [path]);
+
+    return <div>
+        <TableGroup fields={{
+        "Path": { label: nlsHPCC.Path, type: "string", value: value },
+            }} onChange={(id, value) => {
+            setPath (value);
+            }} />
```
Comment (GordonSmith) [inline esp/src/src-react/components/GetValue.tsx:38]: Lint issues

File: esp/src/src-react/routes.tsx [hunk 11]
```diff
@@ -224,6 +223,14 @@ export const routes: RoutesEx = [
             { path: "/bindings/:Name/:Tab", action: (ctx, params) => import("./components/DESDLBindingDetails").then(_ => <_.DESDLBindingDetails name={params.Name as string} tab={params.Tab as string} />) },
         ]
     },
+    {
+        mainNav: ["topology-old"],
+        path: "/daliadmin",
+        children: [
+            { path: "", action: (ctx, params) => import("./components/DaliAdmin").then(_ => <_.DaliAdmin />) },
+            { path: "/:Tab", action: (ctx, params) => import("./components/DaliAdmin").then(_ => <_.DaliAdmin tab={params.Tab as string} />) },
+        ]
```
Comment (GordonSmith) [inline esp/src/src-react/routes.tsx:232]: These routes can be moved to the "topology" section

File: esp/src/src/nls/hpcc.ts [hunk 12]
```diff
@@ -334,6 +336,13 @@ export = {
         FromTime: "From Time",
         FullName: "Full Name",
         Generate: "Generate",
+        GetDFSCSV: "Get DFS CSV",
+        GetDFSMap: "Get DFS Map",
+        GetDFSParents:"Get DFS Parents",
+        GetLogicalFile:"Get Logical File",
+        GetLogicalFilePart: "Get Logical File Part",
+        GetProtectedList: "Get Protected List",
+        GetValue: "Get Value",
```
Comment (GordonSmith) [inline esp/src/src/nls/hpcc.ts:345]: For consistency this PR should drop the "Get" verb from these translations.

File: esp/src/src-react/components/GetDFSMap.tsx [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+
+}
+
+export const GetDFSMap: React.FunctionComponent<GetDFSMapProps> = ({
+
+}) => {
+
+    const [columns, setColumns] = React.useState<IColumn[]>([]);
+    const [items, setItems] = React.useState<any[]>([]);
+    //const [dfsMap, setDfsMap] = React.useState<string>("");
```
Comment (jeclrsg) [inline esp/src/src-react/components/GetDFSMap.tsx:31]: Commented out code that could be deleted.

File: esp/src/src-react/components/GetDFSMap.tsx [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+                        layoutMode={DetailsListLayoutMode.justified}
+                        selectionPreservedOnEmptyClick={true}
+                        styles={{ root: { height, minHeight: height, maxHeight: height } }}
+                    />
+                </div>
+            </div>;
+        }}</SizeMe>}
+    />;
+}; 
+//     >
```
Comment (jeclrsg) [inline esp/src/src-react/components/GetDFSMap.tsx:72]: commented code

File: esp/src/src-react/components/GetLogicalFile.tsx [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+interface GetLogicalFileProps {
+}
+
+export const GetLogicalFile: React.FunctionComponent<GetLogicalFileProps> = ({
+
+}) => {
+
+    const [columns, setColumns] = React.useState<IColumn[]>([]);
+    const [items, setItems] = React.useState<any[]>([]);
+    //const [logicalFile, setLogicalFile] = React.useState<string>("");
```
Comment (jeclrsg) [inline esp/src/src-react/components/GetLogicalFile.tsx:30]: commented code

File: esp/src/src-react/components/GetLogicalFilePart.tsx [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+                return {
+                    key: col,
+                    name: col,
+                    fieldName: col,
+                    minWidth: 100
+                };
+            }));
+            setItems(data);
+        }).catch(err => logger.error(err));
+    }, [fileName]);
```
Comment (jeclrsg) [inline esp/src/src-react/components/GetLogicalFilePart.tsx:46]: There are some lint errors / warnings in this file.

File: esp/src/src-react/components/GetProtectedList.tsx [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+                return {
+                    key: col,
+                    name: col,
+                    fieldName: col,
+                    minWidth: 100
+                };
+            }));
+            setItems(data);
+        }).catch(err => logger.error(err));
+    }, [fileName]);
```
Comment (jeclrsg) [inline esp/src/src-react/components/GetProtectedList.tsx:46]: Lint issues here as well.

**Review by jeclrsg — CHANGES_REQUESTED** [review]
@kunalaswani noticed a couple places that had some commented out code that could be removed. And a couple of files had lint problems. You can run `npm run lint` and it will show things that might not have shown up in the "Problems" tab of VSCode.

**Review by jeclrsg — CHANGES_REQUESTED** [review]

File: esp/src/src-react/components/GetProtectedList.tsx (hunk 17 above)
Comment (jeclrsg) [inline esp/src/src-react/components/GetProtectedList.tsx:46]: Same as in GetLogicalFilePart.tsx, array should include all dependencies, `}, [fileName, callerId]);` Currently, the useCallback only updates when fileName changes, so the GetProtectedList request isn't passing anything to the ESP

File: esp/src/src-react/components/GetProtectedList.tsx [hunk 18]
```diff
@@ -0,0 +1,80 @@
...
+
+    return <HolyGrail
+        header={<span><TableGroup fields={{
+            "FileName": { label: nlsHPCC.FileName, type: "string", value: fileName },
+            "CallerId": { label: nlsHPCC.CallerID, type: "string", value: callerId },
+        }} onChange={(id, value) => {
+            switch (id) {
+                case "filename":
```
Comment (jeclrsg) [inline esp/src/src-react/components/GetProtectedList.tsx:54]: Same here with the casing for these switch conditions, should match the keys in the fields object a couple lines up.

File: esp/src/src-react/components/GetLogicalFilePart.tsx [hunk 19]
```diff
@@ -0,0 +1,80 @@
...
+}
+
+export const GetLogicalFilePart: React.FunctionComponent<GetLogicalFilePartProps> = ({
+
+}) => {
+
+    const [columns, setColumns] = React.useState<IColumn[]>([]);
+    const [items, setItems] = React.useState<any[]>([]);
+    const [fileName, setFileName] = React.useState<string>("");
+    const [partNumber, setPartNumber] = React.useState<number>();
```
Comment (jeclrsg) [inline esp/src/src-react/components/GetLogicalFilePart.tsx:31]: This should probably have a default value, `...useState<number>(1);`, or 0(?), whatever makes sense.

**Review by jeclrsg — CHANGES_REQUESTED** [review]
@kunalaswani Sorry, I missed a couple other things earlier.

File: esp/src/src-react/components/GetLogicalFilePart.tsx [hunk 20]
```diff
@@ -0,0 +1,80 @@
...
+        }).catch(err => logger.error(err));
+    }, [fileName, partNumber]);
+
+    return <HolyGrail
+        header={<span><TableGroup fields={{
+            "FileName": { label: nlsHPCC.FileName, type: "string", value: fileName },
+            "PartNumber": { label: nlsHPCC.PartNumber, type: "number", value: partNumber },
+        }} onChange={(id, value) => {
+            switch (id) {
+                case "fileName":
```
Comment (jeclrsg) [inline esp/src/src-react/components/GetLogicalFilePart.tsx:54]: This problem persists. The casing here should match the keys in the fields object (lines 50 & 51), eg: FileName and PartNumber. I wouldn't think this is the case, but perhaps it's a difference between OSX and Windows? But generally speaking, you can see whether or not the onChange callback function is updating the field correctly when viewing the page.

File: esp/src/src-react/components/GetProtectedList.tsx [hunk 21]
```diff
@@ -0,0 +1,80 @@
...
+        }).catch(err => logger.error(err));
+    }, [fileName, callerId]);
+
+    return <HolyGrail
+        header={<span><TableGroup fields={{
+            "FileName": { label: nlsHPCC.FileName, type: "string", value: fileName },
+            "CallerId": { label: nlsHPCC.CallerID, type: "string", value: callerId },
+        }} onChange={(id, value) => {
+            switch (id) {
+                case "fileName":
```
Comment (jeclrsg) [inline esp/src/src-react/components/GetProtectedList.tsx:54]: Same as in GetLogicalFilePart.tsx, these should match the casing of the field keys, so FileName and CallerId. Alternatively, if you were wanting to have them be consistent with the variable names, you could change the casing on lines 50 & 51 to be:
`"fileName": { label: ... },`
`"callerId": { label: ... }`
The onChange callback is switching on that "id" to determine what to do, so it just needs to match, either way.

**Review by jeclrsg — CHANGES_REQUESTED** [review]
@kunalaswani Those GetLogicalFilePart and GetProtectedList tabs still have that problem with the fields not updating due to the mismatched casing vs the field IDs

**Review by jeclrsg — APPROVED** [review]

File: esp/src/src-react/components/DaliAdmin.tsx [hunk 22]
```diff
@@ -0,0 +1,57 @@
...
+    name?: string;
+    baseDn?: string;
+}
+
+export const DaliAdmin: React.FunctionComponent<DaliAdminProps> = ({
+    filter,
+    tab = "getdfscsv",
+    name,
+    baseDn
+}) => {
```
Comment (GordonSmith) [inline esp/src/src-react/components/DaliAdmin.tsx:27]: Some of these props don't appear to be used?

**Review by GordonSmith — CHANGES_REQUESTED** [review]
Looks like the package-lock.json and package.json are out of sync?

File: esp/src/src-react/components/DaliAdmin.tsx (hunk 22 above)
Comment (kunalaswani) [reply esp/src/src-react/components/DaliAdmin.tsx:24]: unused props removed.

File: esp/src/src-react/components/DaliAdmin.tsx [hunk 23]
```diff
@@ -0,0 +1,54 @@
...
+import { GetProtectedList } from "./GetProtectedList";
+import { GetValue } from "./GetValue";
+import nlsHPCC from "src/nlsHPCC";
+
+interface DaliAdminProps {
+    filter?: object;
+    tab?: string;
+    name?: string;
+    baseDn?: string;
+}
```
Comment (GordonSmith) [inline esp/src/src-react/components/DaliAdmin.tsx:20]: @kunalaswani they should be removed from here as well

**Review by GordonSmith — CHANGES_REQUESTED** [review]

**Review by GordonSmith — APPROVED** [review]

**Review by GordonSmith — CHANGES_REQUESTED** [review]
@kunalaswani looks like your package.json and package-lock.json are out of sync?

On you machine:
* Checkout this PR branch
* Run `npm install && npm audit --fix`
* Amend the commit and push to your repo on GitHub

**Comment by kunalaswani** [comment]
@GordonSmith please review, applied npm update

**Review by GordonSmith — APPROVED** [review]

**Pull request merged**